ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -816,14 +816,29 @@ let checkpoint_in_progress ~__context ~vm =
     (List.map snd (Db.VM.get_current_operations ~__context ~self:vm))
   |> List.mem `checkpoint
 
+let remove_pending_guidance ~__context ~self ~value =
+  debug "Remove guidance [%s] from vm [%s]'s all pending guidances lists"
+    Updateinfo.Guidance.(of_pending_guidance value |> to_string)
+    (Db.VM.get_name_label ~__context ~self) ;
+  Db.VM.remove_pending_guidances ~__context ~self ~value ;
+  Db.VM.remove_pending_guidances_recommended ~__context ~self ~value ;
+  Db.VM.remove_pending_guidances_full ~__context ~self ~value
+
 (** 1. Called on new VMs (clones, imports) and on server start to manually refresh
     the power state, allowed_operations field etc.  Current-operations won't be
     cleaned
     2. Called on update VM when the power state changes *)
 let force_state_reset_keep_current_operations ~__context ~self ~value:state =
   (* First update the power_state. Some operations below indirectly rely on this. *)
   Db.VM.set_power_state ~__context ~self ~value:state ;
+  if state = `Suspended then
+    if not (Db.VM.get_is_a_template ~__context ~self) then
```

Comment:
Why do we need to check for templates? I don't see why pending_guidance needs to be preserved for those?

Reply:
Yes, `pending_guidance` don't needs to be preserved for templates. The if judgement here is to avoid clearing guidance for a template as it is not needed for templates. (And otherwise I think we will also need to update the log in `remove_pending_guidance` to avoid meaningless clearing guidance logs for templates.)

Reply:
Templates are normally running, so this function `force_state_reset` is rarely (never?) called on templates. I'd like the keep the code simple and avoid two many cases, to not make people reading the code wonder what it is for and to avoid unexpected side-effects.

Reply:
> Templates are normally running
Sorry, do you mean template's `power-state` is be running, but I checked their power-state, it is halted, for example:
`uuid ( RO)                : 03d30863-86a2-428e-94c8-88ddfcb8cd29
          name-label ( RW): Red Hat Enterprise Linux 7
    name-description ( RW): To use this template from the CLI, install your VM using vm-install, then set other-config-install-repository to the path to your network repository, e.g. http://<server>/<path> or nfs:server:/<path>


uuid ( RO)                : 7774689b-4ca1-4dea-8545-dddd6b64c17f
          name-label ( RW): Windows 10 (64-bit)
    name-description ( RW): Clones of this template will automatically provision their storage when first booted and then reconfigure themselves with the optimal settings for Windows 10 (64-bit).`

`[root@xrtmia-12-02 ~]# xe template-param-list uuid=7774689b-4ca1-4dea-8545-dddd6b64c17f | grep power-state
                           power-state ( RO): halted
[root@xrtmia-12-02 ~]# xe template-param-list uuid=03d30863-86a2-428e-94c8-88ddfcb8cd29 | grep power-state
                           power-state ( RO): halted
`

And with the `if` check removed, during xapi startup, `remove_pending_guidance` is called for each template, here is the log during xapi startup:
`Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [SUSE Linux Enterprise Server 12 SP4 (64-bit) (deprecated)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [SUSE Li nux Enterprise Server 12 SP4 (64-bit) (deprecated)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Windows Server 2016 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Windows  Server 2016 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Red Hat Enterprise Linux 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Red Hat  Enterprise Linux 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Debian Buster 10]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Debian  Buster 10]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Windows Server 2019 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Windows  Server 2019 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Red Hat Enterprise Linux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Red Hat  Enterprise Linux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Ubuntu Jammy Jellyfish 22.04]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Ubuntu  Jammy Jellyfish 22.04]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Gooroom Platform 2.0]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Gooroom  Platform 2.0]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Ubuntu Focal Fossa 20.04]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Ubuntu  Focal Fossa 20.04]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Oracle Linux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Oracle  Linux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [CentOS 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [CentOS  7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Oracle Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Oracle  Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Windows Server 2022 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Windows  Server 2022 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Debian Bullseye 11]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Debian  Bullseye 11]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Debian Bookworm 12 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Debian  Bookworm 12 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [test 2: no memory, taken with guidance]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [test 2:  no memory, taken with guidance]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Ubuntu Bionic Beaver 18.04 (deprecated)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Ubuntu  Bionic Beaver 18.04 (deprecated)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Scientific Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Scienti fic Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Other install media]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Other i nstall media]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Rocky Linux 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Rocky L inux 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Windows 11 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Windows  11 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [NeoKylin Linux Server 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [NeoKyli n Linux Server 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [CentOS Stream 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [CentOS  Stream 9 (preview)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [SUSE Linux Enterprise 15 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [SUSE Li nux Enterprise 15 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [SUSE Linux Enterprise Server 12 SP5 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [SUSE Li nux Enterprise Server 12 SP5 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Rocky Linux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Rocky L inux 8]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Red Hat Enterprise Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Red Hat  Enterprise Linux 7]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartDeviceModel] from vm  [Windows 10 (64-bit)]'s all pending guidances lists
Jan 18 14:40:28 xrtmia-12-02 xapi: [debug||0 |dbsync (update_env) R:82eafac50a6b|xapi_vm_lifecycle] Remove guidance [RestartVM] from vm [Windows  10 (64-bit)]'s all pending guidances lists
`

Reply:
Sorry I missed the word "not", for "not running" 😅

Reply:
Sure :)
Since this causes confusion, I will remove the `if` check here and update the log in `remove_pending_guidance` to avoid those meaningless log.